packages/api/internal/dns/server.go

```diff
@@ -1,48 +1,81 @@
 package dns
 
 import (
+	"context"
 	"fmt"
-	"log"
 	"net"
 	"strings"
-	"sync"
 
+	redis "github.com/go-redis/redis/v8"
 	resolver "github.com/miekg/dns"
-
-	"github.com/e2b-dev/infra/packages/shared/pkg/smap"
+	"go.uber.org/zap"
 )
 
 const ttl = 0
 
 const defaultRoutingIP = "127.0.0.1"
 
+type FallbackResolverFn = func(sandboxID string) (string, bool)
+
 type DNS struct {
-	mu      sync.Mutex
-	records *smap.Map[string]
+	ctx                context.Context
+	rdb                *redis.Client
+	fallbackResolverFn FallbackResolverFn
+	logger             *zap.SugaredLogger
 }
 
-func New() *DNS {
+func New(ctx context.Context, rdbOpts *redis.Options, fallbackResolverFn FallbackResolverFn, logger *zap.SugaredLogger) *DNS {
 	return &DNS{
-		records: smap.New[string](),
+		ctx:                ctx,
+		rdb:                redis.NewClient(rdbOpts),
+		fallbackResolverFn: fallbackResolverFn,
+		logger:             logger,
 	}
 }
 
-func (d *DNS) Add(sandboxID, ip string) {
-	d.records.Insert(d.hostname(sandboxID), ip)
+func (d *DNS) Add(sandboxID, ip string) error {
+	d.logger.Infof("DNS: Adding entry, sandboxID=%s -> %s", sandboxID, ip)
+	if err := d.rdb.Set(d.ctx, d.dnsKeyFor(sandboxID), ip, 86400).Err(); err != nil {
```

Comment:
Use constant instead
```suggestion
	if err := d.rdb.Set(d.ctx, d.dnsKeyFor(sandboxID), ip, time.Hour * 24).Err(); err != nil {
```

Reply:
👍